### PR TITLE
[Feature]added user model and cooperative model

### PR DESCRIPTION
**Pull request by arj9912**

### CODE REVIEW

File: cooperative/models.py [hunk 1]
```diff
@@ -15,44 +16,102 @@ def __str__(self):
...
-    NATURE_TERM = "term"
-    NATURE_OVERDRAFT = "overdraft"
+    TYPE_TERM = "term"
+    TYPE_OVERDRAFT = "overdraft"
 
-    NATURE_CHOICES = [
-        (NATURE_TERM, "Term"),
-        (NATURE_OVERDRAFT, "Overdraft (OD)"),
+    TYPE_CHOICES = [
+        (TYPE_TERM, "Term"),
```
Comment (RahulRimal) [inline cooperative/models.py:76]: change TYPE_TERM to NATURE_TERM and TYPE_OVERDRAFT to NATURE_OVERDRAFT

File: cooperative/models.py [hunk 2]
```diff
@@ -15,44 +16,102 @@ def __str__(self):
...
-    currently_outstanding = models.DecimalField(max_digits=12, decimal_places=2)
-    total_due = models.DecimalField(max_digits=12, decimal_places=2)
+
+    def __str__(self):
+        return self.account_number
+    
+
+class Installment(BaseModelMixin):
+    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name="installments")
+    due_date = models.DateField(default=now)
```
Comment (RahulRimal) [inline cooperative/models.py:105]: remove default

File: cooperative/models.py [hunk 3]
```diff
@@ -116,3 +175,40 @@ class LoanApplication(BaseModelMixin):
     loan_amount = models.DecimalField(max_digits=10, decimal_places=2)
     finance = models.ForeignKey(Finance, on_delete=models.CASCADE, related_name="loan_applications")
     status = models.CharField(max_length=10, choices=STATUS_CHOICES, default=STATUS_PENDING)
+
+
+
+class Security(BaseModelMixin):
+    LOAN_TYPE_REAL_ESTATE = "real estate"
```
Comment (RahulRimal) [inline cooperative/models.py:182]: remove LOAN_ from and make TYPE_REAL_ESTATE and so on

File: cooperative/models.py [hunk 4]
```diff
@@ -116,3 +175,40 @@ class LoanApplication(BaseModelMixin):
...
+    LOAN_TYPE_FIXED_ASSET = "fixed asset"
+    LOAN_TYPE_HIGHER_PURCHASE = "higher purchase"
+
+    LOAN_TYPE_CHOICES = [
```
Comment (RahulRimal) [inline cooperative/models.py:186]: make LOAN_TYPE_CHOICES to TYPE_CHOICES

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: cooperative/models.py (hunk 1 above)
Comment (arj9912) [reply cooperative/models.py]: done

File: cooperative/models.py (hunk 2 above)
Comment (arj9912) [reply cooperative/models.py]: done

File: cooperative/models.py (hunk 3 above)
Comment (arj9912) [reply cooperative/models.py]: done

File: cooperative/models.py (hunk 4 above)
Comment (arj9912) [reply cooperative/models.py]: done

File: cooperative/serializers.py [hunk 5]
```diff
@@ -73,53 +75,116 @@ class Meta:
...
+            "overdue_amount",
+            "status",
+            "loan_type",
+            "is_closed",
+        ]
+    
+
+
+class InstallmentSerializer(BaseModelSerializerMixin):
+    loan = LoanSerializer(read_only=True)
```
Comment (RahulRimal) [inline cooperative/serializers.py:149]: Remove commented code

File: cooperative/serializers.py [hunk 6]
```diff
@@ -300,3 +400,45 @@ class Meta:
...
+            "description",
+            "ownership_type",
+            "coverage_percentage",
+            "nature_of_charge",
+            "latest_value",
+            "latest_valuation_date",
+        ]
+
+
+class CreateSecuritySerializer(BaseModelSerializerMixin):
```
Comment (RahulRimal) [inline cooperative/serializers.py:422]: Rename to CreateSecurityDepositSerializer

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: cooperative/models.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
 from django.db import models
 from autho.models import User
+from django.utils.timezone import now
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py [hunk 8]
```diff
@@ -16,43 +17,68 @@ def __str__(self):
...
+
+class Installment(BaseModelMixin):
+    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name="installments")
+    due_date = models.DateField()
+    paid_date = models.DateField()
+    total_due = models.DecimalField(max_digits=10, decimal_places=2)
+    over_paid = models.DecimalField(max_digits=10, decimal_places=2)
+    total_outstanding = models.DecimalField(max_digits=10, decimal_places=2)
+
+    def __str__(self):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py [hunk 9]
```diff
@@ -116,3 +142,40 @@ class LoanApplication(BaseModelMixin):
     loan_amount = models.DecimalField(max_digits=10, decimal_places=2)
     finance = models.ForeignKey(Finance, on_delete=models.CASCADE, related_name="loan_applications")
     status = models.CharField(max_length=10, choices=STATUS_CHOICES, default=STATUS_PENDING)
+
+
+
+class Security(BaseModelMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py [hunk 10]
```diff
@@ -116,3 +142,40 @@ class LoanApplication(BaseModelMixin):
...
+    TYPE_REAL_ESTATE = "real state"
+    TYPE_FIXED_ASSET = "fixed asset"
+    TYPE_HIGHER_PURCHASE = "higher purchase"
+
+    TYPE_CHOICES = [
+        (TYPE_REAL_ESTATE, "Real State"),
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py [hunk 11]
```diff
@@ -116,3 +142,40 @@ class LoanApplication(BaseModelMixin):
...
+    type = models.CharField(max_length=20, choices=TYPE_CHOICES, default=TYPE_REAL_ESTATE)
+    description = models.TextField()
+    ownership_type = models.CharField(max_length=20, choices=OWNERSHIP_CHOICES, default=OWNERSHIP_OWN)
+    coverage_percentage = models.DecimalField(max_digits=5, decimal_places=2)
+    nature_of_charge = models.CharField(max_length=20, default=NATURE_FIRST_CHARGE, editable=False)
+    latest_value = models.DecimalField(max_digits=15, decimal_places=2)
+    latest_valuation_date = models.DateField()
+
+    def __str__(self):
+        return self.loan.account_number
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/serializers.py [hunk 12]
```diff
@@ -7,9 +7,11 @@
 from cooperative.models import (
     Company,
     Finance,
+    Installment,
     Loan,
     LoanApplication,
     PersonalGuarantor,
+    Security,
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/serializers.py]: ok done

File: cooperative/serializers.py (hunk 5 above)
Comment (arj9912) [reply cooperative/serializers.py:148]: ok done

File: cooperative/serializers.py (hunk 6 above)
Comment (arj9912) [reply cooperative/serializers.py]: ok done

File: autho/models.py [hunk 13]
```diff
@@ -54,6 +59,8 @@ class User(AbstractBaseUser, PermissionsMixin, BaseModelMixin):
     citizenship_issued_date = models.DateField()
     dob = models.DateField(blank=True, null=True)
     father_name = models.CharField(max_length=255)
+    mother_name = models.CharField(max_length=255)
+    grandfathers_name = models.CharField(max_length=255)
```
Comment (RahulRimal) [inline autho/models.py:63]: change to grandfather_name

File: autho/serializers.py [hunk 14]
```diff
@@ -21,6 +21,8 @@ class Meta:
             "citizenship_issued_date",
             "dob",
             "father_name",
+            "mother_name",
+            "grandfathers_name",
```
Comment (RahulRimal) [inline autho/serializers.py:25]: change to grandfather_name

File: cooperative/models.py [hunk 15]
```diff
@@ -16,43 +16,64 @@ def __str__(self):
...
-    DUE_TYPE_MONTHLY = "m"
-    DUE_TYPE_QUARTERLY = "q"
-    DUE_TYPE_YEARLY = "y"
-
-    INSTALLMENT_DUE_TYPE_CHOICES = [
-        (DUE_TYPE_DAILY, "Daily"),
-        (DUE_TYPE_MONTHLY, "Monthly"),
-        (DUE_TYPE_QUARTERLY, "Quarterly"),
-        (DUE_TYPE_YEARLY, "Yearly"),
+    STATUS_GOOD = "Good"
```
Comment (RahulRimal) [inline cooperative/models.py:20]: change to good

File: cooperative/models.py [hunk 16]
```diff
@@ -16,43 +16,64 @@ def __str__(self):
...
+    STATUS_WATCHLIST = "Watchlist"
+    STATUS_PASS = "Pass"
+    STATUS_NPL = "Non-Performing Loan"
```
Comment (RahulRimal) [inline cooperative/models.py:23]: change to npl

File: cooperative/models.py [hunk 17]
```diff
@@ -16,43 +16,64 @@ def __str__(self):
...
+    STATUS_DOUBTFUL = "Doubtful"
+    STATUS_BAD_DEBT = "Bad Debt"
+
+    STATUS_CHOICES = [
+        (STATUS_GOOD, "Good"),
```
Comment (RahulRimal) [inline cooperative/models.py:28]: use STATUS_GOOD.capitalize() instead of "Good"

File: cooperative/models.py [hunk 18]
```diff
@@ -116,3 +137,40 @@ class LoanApplication(BaseModelMixin):
...
+        (TYPE_REAL_ESTATE, TYPE_REAL_ESTATE.capitalize()),
+        (TYPE_FIXED_ASSET, TYPE_FIXED_ASSET.capitalize()),
+        (TYPE_HIGHER_PURCHASE, TYPE_HIGHER_PURCHASE.capitalize()),
+    ]
+
+    OWNERSHIP_OWN = "own"
+    OWNERSHIP_THIRD_PARTY = "third party"
+
+    OWNERSHIP_CHOICES = [
+        (OWNERSHIP_OWN, "Own"),
```
Comment (RahulRimal) [inline cooperative/models.py:158]: change to OWNERSHIP_WON.capitalize() instead

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: autho/models.py (hunk 13 above)
Comment (arj9912) [reply autho/models.py]: ok done

File: autho/serializers.py (hunk 14 above)
Comment (arj9912) [reply autho/serializers.py]: ok done

File: cooperative/models.py (hunk 15 above)
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py (hunk 16 above)
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py (hunk 17 above)
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py (hunk 18 above)
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py [hunk 19]
```diff
@@ -17,28 +17,28 @@ def __str__(self):
...
-    STATUS_PASS = "Pass"
-    STATUS_NPL = "Non-Performing Loan"
-    STATUS_DOUBTFUL = "Doubtful"
-    STATUS_BAD_DEBT = "Bad Debt"
+    STATUS_GOOD = "good"
+    STATUS_WATCHLIST = "watchlist"
+    STATUS_PASS = "pass"
+    STATUS_NPL = "npl"
+    STATUS_DOUBTFUL = "doubtful"
+    STATUS_BAD_DEBT = "bad Debt"
```
Comment (RahulRimal) [inline cooperative/models.py:25]: make it bad dept

File: cooperative/models.py [hunk 20]
```diff
@@ -17,28 +17,28 @@ def __str__(self):
...
-    STATUS_GOOD = "Good"
-    STATUS_WATCHLIST = "Watchlist"
-    STATUS_PASS = "Pass"
-    STATUS_NPL = "Non-Performing Loan"
-    STATUS_DOUBTFUL = "Doubtful"
-    STATUS_BAD_DEBT = "Bad Debt"
+    STATUS_GOOD = "good"
+    STATUS_WATCHLIST = "watchlist"
+    STATUS_PASS = "pass"
+    STATUS_NPL = "npl"
```
Comment (RahulRimal) [inline cooperative/models.py:23]: add comment saying non performing loan

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: cooperative/models.py (hunk 19 above)
Comment (arj9912) [reply cooperative/models.py]: ok done

File: cooperative/models.py (hunk 20 above)
Comment (arj9912) [reply cooperative/models.py:24]: ok done

**Pull request merged**